### PR TITLE
Remove lsp--extract-line-from-buffer

### PR DESCRIPTION
**Pull request by leungbk**
When lsp-find-references searches a file for references, it causes
lsp--line-char-to-point to be called once for each reference in that
file, which can be time-consuming since each invocation of
lsp--line-char-to-point requires a run
of (save-excursion (save-restriction (widen) (goto-char (point-min)
[...]))). Instead, collect all the references within Emacs with O(1)
passes per file using lsp--xref-make-items, which
uses (save-excursion (save-restriction (widen) (goto-char (point-min)
[...]))) once per file.

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -4601,34 +4601,53 @@ Others: TRIGGER-CHARS"
...
-                            'highlight t line)
-    ;; LINE is nil when FILENAME is not being current visited by any buffer.
-    (xref-make (or line filename)
-               (xref-make-file-location filename
-                                        (1+ (gethash "line" pos-start))
-                                        (gethash "character" pos-start)))))
+      (save-restriction
+        (widen)
+        (goto-char (point-min))
+        (dolist (range ranges)
```
Comment (yyoncho) [inline lsp-mode.el:4623]: is there a reason for preferring `dolist` over `-map`?

Comment (leungbk) [reply lsp-mode.el:4623]: I prefer to use `-map` with projection functions that do not mutate anything outside the scope they define, and to use `dolist` for impure functions.

**Comment by leungbk** [comment]
> Can we unify somehow the logic here with the logic in imenu?

I can't think of a great way to do it.

1. `(save-excursion (save-restriction (widen) [...]` is so commonly used that making it part of a macro does not help with expressiveness.
2. Beyond the fact that the data of interest are shaped differently in imenu and in xref, their respective `lsp-` functions modify `point` differently: `lsp--convert-col-to-points-batch` used by `imenu` can change `point` to a different column within a line, while the analog in this PR does not consider the column number when mutating `point`

**Comment by yyoncho** [comment]
> > Can we unify somehow the logic here with the logic in imenu?
> 
> I can't think of a great way to do it.
> 
> 1. `(save-excursion (save-restriction (widen) [...]` is so commonly used that making it part of a macro does not help with expressiveness.

As part of org-mode integration, I have created the following macro:

``` elisp
(defmacro lsp-save-restriction-and-excursion (&rest form)
  (declare (indent 0) (debug t))
  `(if lsp--already-widened
       (save-excursion ,@form)
     (widen)
     (-let [lsp--already-widened t]
       (save-restriction (save-excursion ,@form)))))
```
which aims to address https://github.com/emacs-lsp/lsp-mode/issues/977. Not sure if it can help for that particular case. 

> 2. Beyond the fact that the data of interest are shaped differently in imenu and in xref, their respective `lsp-` functions modify `point` differently: `lsp--convert-col-to-points-batch` used by `imenu` can change `point` to a different column within a line, while the analog in this PR does not consider the column number when mutating `point`

Isn't it possible to have cur-line as a global variable and use it as a pivot in a macro section? In that section, all position->point conversion will be relative to cur-line?

**Comment by yyoncho** [comment]
One more thing to consider: as part of lsp-mode->org-mode integration we will have to go and adjust all point<>position transformations.

**Comment by leungbk** [comment]
> As part of org-mode integration, I have created the following macro:

I'm fine with using that macro here since you seem to have a use for it elsewhere.

> Isn't it possible to have cur-line as a global variable and use it as a pivot in a macro section? In that section, all position->point conversion will be relative to cur-line?

Are there any benefits to making `cur-line` global beyond shaving off several lines of local declarations? I'm not sure I understand what you mean by pivot.

> One more thing to consider: as part of lsp-mode->org-mode integration we will have to go and adjust all point<>position transformations so every time we introduce section like that this increases the complexity of the code.

Do you have a branch of your org-mode extensions I can look at?

**Comment by yyoncho** [comment]
> Are there any benefits to making `cur-line` global beyond shaving off several lines of local declarations? I'm not sure I understand what you mean by pivot.

The idea is that you could generalize both imenu/xrefs and any sorted position->point conversion. E. g. something like that:

``` elisp
(lsp-begin-sorted-position->point-transformation
   ;; code that uses lsp--position->point-using-cur-line
   )
```

Here it is the org-mode code, but it is still in progress:

https://github.com/yyoncho/lsp-mode/tree/org-mode-lsp-mode

**Comment by leungbk** [comment]
Since there may well end up being streaming support for finding references, it's probably not worth continuing with this PR.

**Comment by yyoncho** [comment]
> Since there may well end up being streaming support for finding references, it's probably not worth continuing with this PR.

IMO streaming is not going to happen. I asked 2 times in the LSP repo if there is server planning to implement streaming and nobody answered...

**Pull request closed without merging**